File: js/src/mouse.ts
/**
 * cloakbrowser-human — Human-like mouse movement and clicking.
 *
 * Trajectory: cubic Bezier with random side bias.
 * Easing: ease-in-out cubic for natural acceleration/deceleration.
 * Wobble: sinusoidal micro-jitter, maximal mid-path, zero at endpoints.
 * Burst pattern: 3-5 points sent without pause, then 8-18ms gap.
 * Overshoot: 15% chance of overshooting target by 3-6px, then correction.
 */

import type { Page, Mouse } from 'playwright-core';
import { HumanConfig, rand, randRange, randIntRange, sleep } from './config.js';

// ---------------------------------------------------------------------------
// Easing
// ---------------------------------------------------------------------------

/** Cubic ease-in-out: slow start, fast middle, slow end. */
function easeInOut(t: number): number {
  return t < 0.5
    ? 4 * t * t * t
    : 1 - Math.pow(-2 * t + 2, 3) / 2;
}

// ---------------------------------------------------------------------------
// Bezier
// ---------------------------------------------------------------------------

interface Point {
  x: number;
  y: number;
}

/** Evaluate a cubic Bezier at parameter t ∈ [0, 1]. */
function bezier(p0: Point, p1: Point, p2: Point, p3: Point, t: number): Point {
  const u = 1 - t;
  const uu = u * u;
  const uuu = uu * u;
  const tt = t * t;
  const ttt = tt * t;
  return {
    x: uuu * p0.x + 3 * uu * t * p1.x + 3 * u * tt * p2.x + ttt * p3.x,
    y: uuu * p0.y + 3 * uu * t * p1.y + 3 * u * tt * p2.y + ttt * p3.y,
  };
}

/** Generate control points with random lateral bias so the curve bends naturally. */
function randomControlPoints(start: Point, end: Point): [Point, Point] {
  const dx = end.x - start.x;
  const dy = end.y - start.y;
  const dist = Math.hypot(dx, dy);
  // Perpendicular direction
  const px = -dy / (dist || 1);
  const py = dx / (dist || 1);
  // Side bias: offset perpendicular to the straight path
  const bias1 = rand(-0.3, 0.3) * dist;
  const bias2 = rand(-0.3, 0.3) * dist;
  return [
    {
      x: start.x + dx * 0.25 + px * bias1,
      y: start.y + dy * 0.25 + py * bias1,
    },
    {
      x: start.x + dx * 0.75 + px * bias2,
      y: start.y + dy * 0.75 + py * bias2,
    },
  ];
}

// ---------------------------------------------------------------------------
// Human mouse movement
// ---------------------------------------------------------------------------

/**
 * Move the mouse from (startX, startY) to (endX, endY) with human-like trajectory.
 * Uses Bezier curves, easing, wobble, and burst patterns.
 */
export async function humanMove(
  page: Page,
  startX: number,
  startY: number,
  endX: number,
  endY: number,
  cfg: HumanConfig,
): Promise<void> {
  const dist = Math.hypot(endX - startX, endY - startY);
  if (dist < 1) return;

  const steps = Math.max(
    cfg.mouse_min_steps,
    Math.min(cfg.mouse_max_steps, Math.round(dist / cfg.mouse_steps_divisor)),
  );

  const start: Point = { x: startX, y: startY };
  const end: Point = { x: endX, y: endY };
  const [cp1, cp2] = randomControlPoints(start, end);

  let burstCounter = 0;
  const burstSize = randIntRange(cfg.mouse_burst_size);

  for (let i = 0; i <= steps; i++) {
    const progress = i / steps;
    const easedT = easeInOut(progress);
    const pt = bezier(start, cp1, cp2, end, easedT);

    // Wobble: sinusoidal jitter, maximal at mid-path, zero at endpoints
    const wobbleAmp = Math.sin(Math.PI * progress) * cfg.mouse_wobble_max;
    const wx = pt.x + (Math.random() - 0.5) * 2 * wobbleAmp;
    const wy = pt.y + (Math.random() - 0.5) * 2 * wobbleAmp;

    await page.mouse.move(Math.round(wx), Math.round(wy));

    // Burst pattern: send burstSize points without pause, then pause
    burstCounter++;
    if (burstCounter >= burstSize && i < steps) {
      await sleep(randRange(cfg.mouse_burst_pause));
      burstCounter = 0;
    }
  }

  // Overshoot: 15% chance the cursor overshoots and corrects
  if (Math.random() < cfg.mouse_overshoot_chance) {
    const overshootDist = randRange(cfg.mouse_overshoot_px);
    const angle = Math.atan2(endY - startY, endX - startX);
    const ovX = Math.round(endX + Math.cos(angle) * overshootDist);
    const ovY = Math.round(endY + Math.sin(angle) * overshootDist);
    await page.mouse.move(ovX, ovY);
    await sleep(rand(30, 70));
    // Correct back to target ±2px
    const corrX = Math.round(endX + (Math.random() - 0.5) * 4);
    const corrY = Math.round(endY + (Math.random() - 0.5) * 4);
    await page.mouse.move(corrX, corrY);
  }
}

// ---------------------------------------------------------------------------
// Human click
// ---------------------------------------------------------------------------

/**
 * Determine the click target point within an element's bounding box.
 * - Input fields: left third (5-30% width, 30-70% height)
 * - Buttons/other: near center (±30% width, ±30% height)
 */
export function clickTarget(
  box: { x: number; y: number; width: number; height: number },
  isInput: boolean,
  cfg: HumanConfig,
): Point {
  if (isInput) {
    const xFrac = randRange(cfg.click_input_x_range);
    const yFrac = rand(0.30, 0.70);
    return {
      x: Math.round(box.x + box.width * xFrac),
      y: Math.round(box.y + box.height * yFrac),
    };
  }
  // Button or generic element — near center
  const xFrac = rand(0.35, 0.65);
  const yFrac = rand(0.35, 0.65);
  return {
    x: Math.round(box.x + box.width * xFrac),
    y: Math.round(box.y + box.height * yFrac),
  };
}

/**
 * Perform a human-like click: mousedown → hold → mouseup.
 * The mouse must already be at the target position.
 */
export async function humanClick(
  page: Page,
  isInput: boolean,
  cfg: HumanConfig,
): Promise<void> {
  // Pre-click aiming delay
  const aimDelay = isInput
    ? randRange(cfg.click_aim_delay_input)
    : randRange(cfg.click_aim_delay_button);
  await sleep(aimDelay);

  // Click: down → hold → up
  const holdTime = isInput
    ? randRange(cfg.click_hold_input)
    : randRange(cfg.click_hold_button);
  await page.mouse.down();
  await sleep(holdTime);
  await page.mouse.up();
}

// ---------------------------------------------------------------------------
// Human idle / drift
// ---------------------------------------------------------------------------

/**
 * Idle drift: cursor stays roughly in place with tiny random movements.
 * Used when the "user" is reading or waiting.
 *
 * @param seconds - approximate duration of idle period
 * @param cx - current cursor X
 * @param cy - current cursor Y
 */
export async function humanIdle(
  page: Page,
  seconds: number,
  cx: number,
  cy: number,
  cfg: HumanConfig,
): Promise<void> {
  const endTime = Date.now() + seconds * 1000;
  let x = cx;
  let y = cy;
  while (Date.now() < endTime) {
    const dx = (Math.random() - 0.5) * 2 * cfg.idle_drift_px;
    const dy = (Math.random() - 0.5) * 2 * cfg.idle_drift_px;
    x += dx;
    y += dy;
    await page.mouse.move(Math.round(x), Math.round(y));
    await sleep(randRange(cfg.idle_pause_range));
  }
}

File: python/cloakbrowser_human/mouse.py
"""cloakbrowser-human — Human-like mouse movement and clicking.

Trajectory: cubic Bezier with random side bias.
Easing: ease-in-out cubic for natural acceleration/deceleration.
Wobble: sinusoidal micro-jitter, maximal mid-path, zero at endpoints.
Burst pattern: 3-5 points sent without pause, then 8-18ms gap.
Overshoot: 15% chance of overshooting target by 3-6px, then correction.
"""

from __future__ import annotations

import math
import random
from typing import TYPE_CHECKING, Tuple

from .config import HumanConfig, rand, rand_range, rand_int_range, sleep_ms

if TYPE_CHECKING:
    from patchright.sync_api import Page


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------

class Point:
    __slots__ = ("x", "y")

    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y


# ---------------------------------------------------------------------------
# Easing
# ---------------------------------------------------------------------------

def _ease_in_out(t: float) -> float:
    """Cubic ease-in-out: slow start, fast middle, slow end."""
    if t < 0.5:
        return 4 * t * t * t
    return 1 - pow(-2 * t + 2, 3) / 2


# ---------------------------------------------------------------------------
# Bezier
# ---------------------------------------------------------------------------

def _bezier(p0: Point, p1: Point, p2: Point, p3: Point, t: float) -> Point:
    """Evaluate a cubic Bezier at parameter t in [0, 1]."""
    u = 1 - t
    uu = u * u
    uuu = uu * u
    tt = t * t
    ttt = tt * t
    return Point(
        uuu * p0.x + 3 * uu * t * p1.x + 3 * u * tt * p2.x + ttt * p3.x,
        uuu * p0.y + 3 * uu * t * p1.y + 3 * u * tt * p2.y + ttt * p3.y,
    )


def _random_control_points(start: Point, end: Point) -> Tuple[Point, Point]:
    """Generate control points with random lateral bias."""
    dx = end.x - start.x
    dy = end.y - start.y
    dist = math.hypot(dx, dy) or 1
    # Perpendicular direction
    px = -dy / dist
    py = dx / dist
    bias1 = rand(-0.3, 0.3) * dist
    bias2 = rand(-0.3, 0.3) * dist
    return (
        Point(start.x + dx * 0.25 + px * bias1, start.y + dy * 0.25 + py * bias1),
        Point(start.x + dx * 0.75 + px * bias2, start.y + dy * 0.75 + py * bias2),
    )


# ---------------------------------------------------------------------------
# Human mouse movement
# ---------------------------------------------------------------------------

def human_move(
    page: Page,
    start_x: float,
    start_y: float,
    end_x: float,
    end_y: float,
    cfg: HumanConfig,
) -> None:
    """Move the mouse from (start_x, start_y) to (end_x, end_y) with human-like trajectory."""
    dist = math.hypot(end_x - start_x, end_y - start_y)
    if dist < 1:
        return

    steps = max(
        cfg.mouse_min_steps,
        min(cfg.mouse_max_steps, round(dist / cfg.mouse_steps_divisor)),
    )

    start = Point(start_x, start_y)
    end = Point(end_x, end_y)
    cp1, cp2 = _random_control_points(start, end)

    burst_counter = 0
    burst_size = rand_int_range(cfg.mouse_burst_size)

    for i in range(steps + 1):
        progress = i / steps
        eased_t = _ease_in_out(progress)
        pt = _bezier(start, cp1, cp2, end, eased_t)

        # Wobble: sinusoidal jitter, maximal at mid-path, zero at endpoints
        wobble_amp = math.sin(math.pi * progress) * cfg.mouse_wobble_max
        wx = pt.x + (random.random() - 0.5) * 2 * wobble_amp
        wy = pt.y + (random.random() - 0.5) * 2 * wobble_amp

        page.mouse.move(round(wx), round(wy))

        # Burst pattern
        burst_counter += 1
        if burst_counter >= burst_size and i < steps:
            sleep_ms(rand_range(cfg.mouse_burst_pause))
            burst_counter = 0

    # Overshoot: chance the cursor overshoots and corrects
    if random.random() < cfg.mouse_overshoot_chance:
        overshoot_dist = rand_range(cfg.mouse_overshoot_px)
        angle = math.atan2(end_y - start_y, end_x - start_x)
        ov_x = round(end_x + math.cos(angle) * overshoot_dist)
        ov_y = round(end_y + math.sin(angle) * overshoot_dist)
        page.mouse.move(ov_x, ov_y)
        sleep_ms(rand(30, 70))
        # Correct back to target +/-2px
        corr_x = round(end_x + (random.random() - 0.5) * 4)
        corr_y = round(end_y + (random.random() - 0.5) * 4)
        page.mouse.move(corr_x, corr_y)


# ---------------------------------------------------------------------------
# Click target calculation
# ---------------------------------------------------------------------------

def click_target(
    box: dict,
    is_input: bool,
    cfg: HumanConfig,
) -> Point:
    """Determine click target point within an element's bounding box.

    Input fields: left third (5-30% width, 30-70% height).
    Buttons/other: near center (+/-30%).
    """
    if is_input:
        x_frac = rand_range(cfg.click_input_x_range)
        y_frac = rand(0.30, 0.70)
    else:
        x_frac = rand(0.35, 0.65)
        y_frac = rand(0.35, 0.65)
    return Point(
        round(box["x"] + box["width"] * x_frac),
        round(box["y"] + box["height"] * y_frac),
    )


# ---------------------------------------------------------------------------
# Human click
# ---------------------------------------------------------------------------

def human_click(page: Page, is_input: bool, cfg: HumanConfig) -> None:
    """Perform a human-like click: mousedown -> hold -> mouseup."""
    # Pre-click aiming delay
    aim_delay = (
        rand_range(cfg.click_aim_delay_input) if is_input
        else rand_range(cfg.click_aim_delay_button)
    )
    sleep_ms(aim_delay)

    # Click: down -> hold -> up
    hold_time = (
        rand_range(cfg.click_hold_input) if is_input
        else rand_range(cfg.click_hold_button)
    )
    page.mouse.down()
    sleep_ms(hold_time)
    page.mouse.up()


# ---------------------------------------------------------------------------
# Human idle / drift
# ---------------------------------------------------------------------------

def human_idle(
    page: Page,
    seconds: float,
    cx: float,
    cy: float,
    cfg: HumanConfig,
) -> None:
    """Idle drift: cursor stays roughly in place with tiny random movements.

    Args:
        seconds: Approximate duration of idle period.
        cx: Current cursor X position.
        cy: Current cursor Y position.
    """
    import time as _time

    end_time = _time.monotonic() + seconds
    x, y = cx, cy
    while _time.monotonic() < end_time:
        dx = (random.random() - 0.5) * 2 * cfg.idle_drift_px
        dy = (random.random() - 0.5) * 2 * cfg.idle_drift_px
        x += dx
        y += dy
        page.mouse.move(round(x), round(y))
        sleep_ms(rand_range(cfg.idle_pause_range))
